File: 0973/solution.java
import java.util.PriorityQueue;

class Solution {
   public int[][] kClosest(int[][] points, int k) {
      PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> {
         int A = a[0] * a[0] + a[1] * a[1];
         int B = b[0] * b[0] + b[1] * b[1];
         return B - A;
      });
      for (int[] point : points) {
         heap.add(point);
         if (heap.size() > k) {
            heap.poll();
         }
      }
      int[][] res = new int[k][2];
      int i = 0;
      while (!heap.isEmpty()) {
         res[i++] = heap.poll();
      }
      return res;
   }
}

File: 0973/solution.rs
use std::collections::BinaryHeap;


impl Solution {
    pub fn k_closest(points: Vec<Vec<i32>>, k: i32) -> Vec<Vec<i32>> {
        let k = k as usize;
        let mut heap: BinaryHeap<(i32, usize)> = BinaryHeap::new();
        
        for (i, point) in points.iter().enumerate() {
            let dist = point[0].pow(2) + point[1].pow(2);
            heap.push((dist, i));
            if heap.len() > k {
                heap.pop();
            }
        }
        
        heap.into_iter()
            .map(|(_, index)| points[index].clone())
            .collect()
    }
}
